server/src-lib/Hasura/Tracing.hs

```diff
@@ -0,0 +1,237 @@
+  -> TraceT m a
+  -> n b
+interpTraceT f (TraceT rwma) = do
+  ctx <- currentContext
+  rep <- currentReporter
+  (b, meta) <- f (runWriterT (runReaderT rwma (ctx, rep)))
+  attachMetadata meta
+  pure b
+
+-- | If the underlying monad can report trace data, then 'TraceT' will
+-- collect it and hand it off to that reporter.
+instance MonadIO m => MonadTrace (TraceT m) where
+  trace name ma = TraceT . ReaderT $ \(ctx, rep) -> do
+    spanId <- liftIO (Rand.randomIO :: IO Word64)
+    let subCtx = ctx { tcCurrentSpan = spanId
+                     , tcCurrentParent = Just (tcCurrentSpan ctx)
+                     }
+    lift . runReporter rep subCtx name . runWriterT $ runReaderT (unTraceT ma) (subCtx, rep)
+
+  currentContext = TraceT (asks fst)
+  
+  currentReporter = TraceT (asks snd)
+
+  attachMetadata = TraceT . tell
+
+instance MonadTrace m => MonadTrace (ReaderT r m) where
+  trace = mapReaderT . trace
+  currentContext = lift currentContext
+  currentReporter = lift currentReporter
+  attachMetadata = lift . attachMetadata
+
+instance MonadTrace m => MonadTrace (StateT e m) where
+  trace = mapStateT . trace
+  currentContext = lift currentContext
+  currentReporter = lift currentReporter
+  attachMetadata = lift . attachMetadata
+
+instance MonadTrace m => MonadTrace (ExceptT e m) where
+  trace = mapExceptT . trace
+  currentContext = lift currentContext
+  currentReporter = lift currentReporter
+  attachMetadata = lift . attachMetadata
+
+-- | A HTTP request, which can be modified before execution.
+data SuspendedRequest m a = SuspendedRequest HTTP.Request (HTTP.Request -> m a)
+
+-- | Extract the trace and parent span headers from a HTTP request
+-- and create a new 'TraceContext'. The new context will contain
+-- a fresh span ID, and the provided span ID will be assigned as
+-- the immediate parent span.
+extractHttpContext :: [HTTP.Header] -> IO (Maybe TraceContext)
+extractHttpContext hdrs = do
+  freshSpanId <- liftIO Rand.randomIO
+  pure $ TraceContext
+    <$> (HTTP.parseHeaderMaybe =<< lookup "X-Hasura-TraceId" hdrs)
+    <*> pure freshSpanId
+    <*> pure (HTTP.parseHeaderMaybe =<< lookup "X-Hasura-SpanId" hdrs)
+
+traceHttpRequest
+  :: MonadTrace m
+  => Text
+  -- ^ human-readable name for this block of code
+  -> m (SuspendedRequest m a)
+  -- ^ an action which yields the request about to be executed and suspends
+  -- before actually executing it
+  -> m a
+traceHttpRequest name f = trace name do
+  SuspendedRequest req next <- f
+  let reqBytes = case HTTP.requestBody req of
+        HTTP.RequestBodyBS bs -> Just (fromIntegral (BS.length bs))
+        HTTP.RequestBodyLBS bs -> Just (BL.length bs)
+        HTTP.RequestBodyBuilder len _ -> Just len
+        HTTP.RequestBodyStream len _ -> Just len
+        _ -> Nothing
+  for_ reqBytes \b ->
+    attachMetadata [("request_body_bytes", fromString (show b))]
+  TraceContext{..} <- currentContext
+  let tracingHeaders =
```

Comment:
We might want to consider using the B3 spec from OpenZipkin here https://github.com/openzipkin/b3-propagation

Reply:
(I don't think this should block a merge though, it's easily changed later)